### PR TITLE
Capi: Ngraph engine

### PR DESCRIPTION
**Pull request by mozga-intel**
Enable the capi pass for the nGraph bridge. BERT for an inference.

### CODE REVIEW

**Comment by mozga-intel** [comment]
@tensor-tang @luotao1 Can you look at this PR?

**Comment by mozga-intel** [comment]
@tensor-tang @luotao1 Can you look at this PR, this pull-request is needed to support BERT model?

File: paddle/fluid/framework/ir/CMakeLists.txt [hunk 1]
```diff
@@ -92,6 +92,14 @@ if(WITH_MKLDNN)
     pass_library(cpu_quantize_squash_pass inference mkldnn)
 endif()
 
+if(WITH_NGRAPH)
+    cc_library(ngraph_subgraph_pass SRCS ngraph_subgraph_pass.cc DEPS ngraph_bridge
+      analysis_helper subgraph_detector graph_pattern_detector pass fuse_pass_base ${op_library_DEPS})
+    set(pass_file ${PADDLE_BINARY_DIR}/paddle/fluid/inference/api/paddle_inference_pass.h)
+    file(APPEND ${pass_file} "USE_PASS(ngraph_subgraph_pass);\n")
+    set(INFER_IR_PASSES ${INFER_IR_PASSES} ngraph_subgraph_pass CACHE INTERNAL "")
```
Comment (luotao1) [inline paddle/fluid/framework/ir/CMakeLists.txt:103]: Why not use 
```
pass_library(ngraph_subgraph_pass inference ngraph)
```

File: paddle/fluid/inference/tests/api/analyzer_resnet50_tester.cc [hunk 2]
```diff
@@ -42,13 +42,17 @@ void SetOptimConfig(AnalysisConfig *cfg) {
...
+void profile(bool use_mkldnn = false, bool use_ngraph = false) {
   AnalysisConfig cfg;
   SetConfig(&cfg);
 
   if (use_mkldnn) {
     cfg.EnableMKLDNN();
   }
+
+  if (use_ngraph) {
+    cfg.EnableMKLDNN();
```
Comment (luotao1) [inline paddle/fluid/inference/tests/api/analyzer_resnet50_tester.cc]: Here is `cfg. EnableNgraph();`

File: paddle/fluid/operators/ngraph/ops/batch_norm_op.h [hunk 3]
```diff
@@ -44,7 +44,8 @@ void BuildBatchNormNode(
   auto x = paddle::platform::GetInputNode(op, "X", ngb_node_map);
```
Comment (luotao1) [inline paddle/fluid/operators/ngraph/ops/batch_norm_op.h]: Why do you update `batch_norm_op.h` and `conv2d_op.h` here? They are not related to this PR.

**Comment by luotao1** [comment]
@Superjomn Could you help review inference api related interface: `paddle_analysis_config.h`, `analysis_config.cc`?

File: paddle/fluid/operators/ngraph/ops/batch_norm_op.h (hunk 3 above)
Comment (mozga-intel) [reply paddle/fluid/operators/ngraph/ops/batch_norm_op.h]: On the whole, these changes are necessary for the CAPI tests - to pass the tests. I agree with you that it is not associated with this PR, but these changes are really small. If you want I can move it to separate PR - but it will generate additional time.

File: paddle/fluid/framework/ir/CMakeLists.txt (hunk 1 above)
Comment (mozga-intel) [reply paddle/fluid/framework/ir/CMakeLists.txt:103]: Yes, your suggestion seems to be fine, Can I take this suggestion into consideration, and do it in the next pull-request?

File: paddle/fluid/inference/tests/api/analyzer_resnet50_tester.cc (hunk 2 above)
Comment (mozga-intel) [reply paddle/fluid/inference/tests/api/analyzer_resnet50_tester.cc]: Done.

File: paddle/fluid/operators/ngraph/ops/batch_norm_op.h (hunk 3 above)
Comment (baojun-nervana) [reply paddle/fluid/operators/ngraph/ops/batch_norm_op.h]: depthwise_conv can be done in a separate PR (https://github.com/PaddlePaddle/Paddle/pull/17454). The batchnorm update is to pass the test due to the BN fast math issue. That is the easiest way to make it align with paddle precision.

File: paddle/fluid/framework/ir/CMakeLists.txt (hunk 1 above)
Comment (baojun-nervana) [reply paddle/fluid/framework/ir/CMakeLists.txt:103]: it seems it won't work due to some dependency

File: paddle/fluid/framework/ir/ngraph_subgraph_pass.cc [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+namespace paddle {
+namespace framework {
+namespace ir {
+
+namespace ANAT = paddle::inference::analysis;
+
+std::string GenerateEngineKey(const std::set<std::string> &engine_inputs,
+                              const std::set<std::string> &engine_outputs,
+                              const std::string &size) {
+  std::string engine_hash_key = "";
```
Comment (tensor-tang) [inline paddle/fluid/framework/ir/ngraph_subgraph_pass.cc:38]: any unit test to make sure the key is unique?

File: paddle/fluid/framework/ir/ngraph_subgraph_pass.cc [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+}
+
+void NgraphSubgraphPass::CreateNgraphEngineOp(framework::ir::Node *node,
+                                              Graph *graph) const {
+  auto *op_desc = node->Op();
+  auto &subgraph = *ANAT::Agent(node).subgraph();
+  PADDLE_ENFORCE(!subgraph.empty());
+
+  framework::ProgramDesc *program_desc =
+      Get<framework::ProgramDesc *>("program");
```
Comment (tensor-tang) [inline paddle/fluid/framework/ir/ngraph_subgraph_pass.cc:99]: Is that fine to us directly modify the program desc @Superjomn @luotao1

File: paddle/fluid/framework/ir/ngraph_subgraph_pass.cc (hunk 4 above)
Comment (baojun-nervana) [reply paddle/fluid/framework/ir/ngraph_subgraph_pass.cc:38]: since capi for inference is known to be identical program, the key is unimportant. If necessary, we can serialize the prgramdesc and make a hash key.

File: paddle/fluid/framework/ir/ngraph_subgraph_pass.cc (hunk 5 above)
Comment (baojun-nervana) [reply paddle/fluid/framework/ir/ngraph_subgraph_pass.cc:99]: I was following tensorrt subgraph approach. https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/inference/analysis/ir_passes/tensorrt_subgraph_pass.cc#L97

File: paddle/fluid/framework/ir/ngraph_subgraph_pass.cc (hunk 4 above)
Comment (tensor-tang) [reply paddle/fluid/framework/ir/ngraph_subgraph_pass.cc:38]: ok

**Review by tensor-tang — COMMENTED** [review]
OK  thanks @mozga-intel @baojun-nervana .
Please fix  the conflicts

**Comment by mozga-intel** [comment]
@tensor-tang @luotao1 Thanks, The conflicts are resolved.

File: paddle/fluid/framework/ir/CMakeLists.txt (hunk 1 above)
Comment (luotao1) [reply paddle/fluid/framework/ir/CMakeLists.txt:103]: > Can I take this suggestion into consideration, and do it in the next pull-request?

Yes, please do it in the next PR.

**Review by luotao1 — APPROVED** [review]
